### PR TITLE
Fixes pack #7

### PR DESCRIPTION
**Pull request by laerreal**
Fixes for:
- widgets,
- Qemu log visualizer,
- source code model,
- debug module.

### v3
- fixup tab in `VarTreeview`
- QLV:
    - work on coding style
    - `__str__` for `InInstr` and `QTrace`
    - extend/fixup comments
    - highlight first instruction in TB
    - fixup typo `is_trace(l0)`

### v2
- use `Thread.name`
- In `VarTreeview.__init__` copy values list and replace items instead of
list building by `insert`.
- QLV:
    - close `log_stream`
    - `feed_lines` is added in its own patch
    - fixup comment about Trace record format
    - `EOL`: end of log marker
    - fixup errors in `QEMULog.feed`
- `__main__` indicates `EOL` after desired lines amount are fed.

### CODE REVIEW

File: debug/runtime.py [hunk 1]
```diff
@@ -150,6 +150,7 @@ def run():
                 print_exc()
 
         t = Thread(target = run)
+        t.setName("RSP client")
```
Comment (BDanAnd) [inline debug/runtime.py]: Maybe better `t.name = "RSP client"`?
https://docs.python.org/2/library/threading.html#threading.Thread.name
https://docs.python.org/3/library/threading.html#threading.Thread.name

File: qlv.py [hunk 2]
```diff
@@ -107,22 +107,37 @@ class QTrace(object):
...
+                    break
             else:
                 break
+            try:
+                addr_val = int("".join(addr), base = 16)
+            except ValueError:
+                continue
+            addrs.append(addr_val)
 
+        # addrs[0] is `cs_base`, addrs[1] is `pc` and addrs[2] is `flags`
```
Comment (BDanAnd) [inline qlv.py]: You don’t use `flags` but it seems to me that there is a potential problem.
From the QEMU source: ```"[" TARGET_FMT_lx "/" TARGET_FMT_lx "/%#x]"```.
From format identifier help: ```%#x (Hex)   0x prefix added to non-zero values```.
For non-zero flags addrs[2] is 0 and addrs[3] is really `flags` value.

File: qlv.py [hunk 3]
```diff
@@ -363,47 +370,58 @@ def new_unrecognized(self, l):
...
 
     def feed_file(self, f):
         # iterate file line by line
-        self.feed(iter(f))
+        feeder =  self.feeder
+        for l in f:
+            feeder.send(l)
+        feeder.send(None)
+
+    def feed_lines(self, f, limit = 1000):
```
Comment (BDanAnd) [inline qlv.py:458]: Maybe add this function in 9619a02b3459413427757e477e8abf9fb0b33f67?

File: qlv.py [hunk 4]
```diff
@@ -363,47 +370,58 @@ def new_unrecognized(self, l):
...
+        feeder =  self.feeder
+        for last, l in izip(xrange(limit), f):
+            feeder.send(l)
+        if last < limit - 1:
+            feeder.send(None)
```
Comment (BDanAnd) [inline qlv.py]: Is it not a problem that `None` is not sent when `last = limit - 1`?

File: qlv.py [hunk 5]
```diff
@@ -446,7 +452,8 @@ def feed_file_by_name(self, file_name):
     print("Reading " + qlogFN)
 
     qlog = QEMULog()
-    qlog.feed_file_by_name(qlogFN)
+    log_stream = open(qlogFN, "r")
+    qlog.feed_lines(log_stream, int(args.l))
```
Comment (BDanAnd) [inline qlv.py]: File does not close.

File: widgets/var_widgets.py [hunk 6]
```diff
@@ -272,17 +272,17 @@ def insert(self, *a, **kw):
...
-                values = (values,)
+            kw["values"] = values = []
 
-            for col, v in enumerate(values):
+            for col, v in enumerate(old_values):
                 if isinstance(v, variables):
                     to_track.append((col, v))
-                    values[col] = v.get()
+                    v = v.get()
+                values.insert(col, v)
```
Comment (BDanAnd) [inline widgets/var_widgets.py]: Maybe:
```
        try:
	    kw["values"] = values = list(kw.pop("values"))
        except KeyError:
            pass
        else:
            for col, v in enumerate(values):
                if isinstance(v, variables):
                    to_track.append((col, v))
                    values[col] = v.get()
```

Comment (laerreal) [reply widgets/var_widgets.py]: Changing list during iteration may raise an exception. Did you tried this snippet?

File: debug/runtime.py (hunk 1 above)
Comment (laerreal) [reply debug/runtime.py]: a newfag is detected ;)

File: qlv.py [hunk 7]
```diff
@@ -363,47 +370,58 @@ def new_unrecognized(self, l):
...
+        l0 = yield
+        while l0 is not None:
+            if is_in_asm(l0):
+                in_asm = []
+                l1 = yield
+                while l1 is not None:
+                    if is_in_asm_instr(l1):
+                        in_asm.append(l1)
                     else:
-                        l0 = None
```
Comment (BDanAnd) [inline qlv.py:360]: This assignment is needed.

File: qlv.py [hunk 8]
```diff
@@ -363,47 +370,58 @@ def new_unrecognized(self, l):
...
+            if is_trace(l0):
+                trace = [l0]
 
-                    for l1 in reader:
-                        if is_trace(l1) or is_linking(l1) or is_in_asm(l1):
-                            l0 = l1
-                            break
-                        trace.append(l1)
-                    else:
-                        l0 = None
```
Comment (BDanAnd) [inline qlv.py:374]: This assignment is needed.

File: qlv.py [hunk 9]
```diff
@@ -363,47 +370,58 @@ def new_unrecognized(self, l):
...
+                l1 = yield
+                while l1 is not None:
+                    if is_linking(l1) or is_in_asm(l1):
+                        l0 = l1
+                        break
+                    l1 = yield
+                    trace.append(l1)
```
Comment (BDanAnd) [inline qlv.py]: Rearrange the lines. The first `l1` value is lost.

File: qlv.py (hunk 4 above)
Comment (laerreal) [reply qlv.py]: It's as expected.

**Comment by laerreal** [comment]
v2 is ready!

File: qlv.py [hunk 10]
```diff
@@ -364,47 +377,62 @@ def new_unrecognized(self, l):
...
+                self.new_linking(l0)
+                l0 = yield
+                continue
 
-                self.new_unrecognized(l0)
-                break
+            self.new_unrecognized(l0)
+            l0 = yield
+
+        # StopIteration` should not be raised because `send` method is used
```
Comment (BDanAnd) [inline qlv.py]: No opening `?

File: qlv.py [hunk 11]
```diff
@@ -107,22 +107,38 @@ class QTrace(object):
...
             else:
                 break
+            try:
+                addr_val = int("".join(addr), base = 16)
+            except ValueError:
+                continue
+            addrs.append(addr_val)
 
+        # addrs[0] is `cs_base`, addrs[1] is `pc`,
+        # addrs[2] is 0 from "0x" prefix of flags, addrs[3] is `flags`
```
Comment (BDanAnd) [inline qlv.py]:
```
printf("%#x %#x", 0, 1);
stdout: 0 0x1
```

File: qlv.py [hunk 12]
```diff
@@ -364,47 +377,62 @@ def new_unrecognized(self, l):
...
-                            l0 = l1
-                            break
-                        trace.append(l1)
-                    else:
-                        l0 = None
+                l1 = yield
+                while l1 is not EOL:
+                    # Traces are following one by one: user did not
+                    # passed other flags to -d.
+                    if is_trace(l0) or is_linking(l1) or is_in_asm(l1):
```
Comment (BDanAnd) [inline qlv.py]: Maybe `is_trace(l1)`?

File: widgets/var_widgets.py [hunk 13]
```diff
@@ -272,13 +272,10 @@ def insert(self, *a, **kw):
         to_track = []
 
         try:
-            values = kw["values"]
+	        kw["values"] = values = list(kw.pop("values"))
```
Comment (BDanAnd) [inline widgets/var_widgets.py]: Sorry. The tab was wrapped in code snippet.

File: widgets/var_widgets.py (hunk 6 above)
Comment (BDanAnd) [reply widgets/var_widgets.py]: I ran the GUI and received no exceptions.

**Comment by laerreal** [comment]
v3 is ready!

**Pull request merged**